src/edge_proxy/environments.py

```diff
@@ -142,11 +142,41 @@ def get_environment(self, client_side_key: str) -> dict[str, typing.Any]:
             return environment_document
         raise FlagsmithUnknownKeyError(client_side_key)
 
-    async def _fetch_document(self, server_side_key: str) -> dict[str, typing.Any]:
+    async def _fetch_document(
+        self, key_pair: EnvironmentKeyPair
+    ) -> dict[str, typing.Any]:
+        headers = {
+            "X-Environment-Key": key_pair.server_side_key,
+        }
+        environment_document = self.cache.get_environment(
+            environment_api_key=key_pair.client_side_key
+        )
+        if environment_document:
+            updated_at: str = environment_document.get("updated_at")
+            if updated_at:
+                try:
+                    epoch_seconds = datetime.fromisoformat(updated_at).timestamp()
+                    # Same implementation as https://docs.djangoproject.com/en/4.2/ref/utils/#django.utils.http.http_date
+                    headers["If-Modified-Since"] = formatdate(
+                        epoch_seconds, usegmt=True
+                    )
+                except ValueError:
+                    logger.warning(
+                        f"failed to parse updated_at, environment={key_pair.client_side_key} updated_at={updated_at}"
+                    )
+            else:
+                logger.warning(
+                    f"received environment with no updated_at: {key_pair.client_side_key}"
+                )
         response = await self._client.get(
             url=f"{self.settings.api_url}/environment-document/",
-            headers={"X-Environment-Key": server_side_key},
+            headers=headers,
         )
+        if response.status_code == starlette.status.HTTP_304_NOT_MODIFIED:
+            if not environment_document:
```

Comment:
I think let's do:

```
assert environment_document, f"GET /environment-document returned 304 without a cached document. environment={key_pair.client_side_key}"
```

Reply:
Added in https://github.com/Flagsmith/edge-proxy/pull/159/commits/c1e26fccdcffee6b62184937bb301c634f15e7bd